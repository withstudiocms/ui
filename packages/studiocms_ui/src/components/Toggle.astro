---
import type { StudioCMSColorway } from '../utils/colors';
import { generateID } from '../utils/generateID';

interface Props {
	label: string;
	size?: 'sm' | 'md' | 'lg';
	color?: StudioCMSColorway;
	defaultChecked?: boolean;
	disabled?: boolean;
	name?: string;
	isRequired?: boolean;
};

const {
	size = 'md',
	color = 'default',
	defaultChecked,
	disabled,
	name = generateID('checkbox'),
	label,
	isRequired,
} = Astro.props;
---
<label
  class="sui-toggle-label"
  for={name}
  class:list={[
    disabled && "disabled",
    color,
    size,
  ]}
>
  <div class="sui-toggle-container">
    <div class="sui-toggle-switch" />
    <input
      type="checkbox"
      name={name}
      id={name}
      checked={defaultChecked}
      disabled={disabled}
      required={isRequired}
      class="sui-checkbox"
    />
  </div>
  <span>
    {label} <span class="req-star">{isRequired && "*"}</span>
  </span>
</label>
<style>
  .sui-toggle-label {
    display: flex;
    flex-direction: row;
    align-items: center;
    gap: .5rem;
    position: relative;
    margin: .25rem 0;
  }

  .sui-toggle-label.disabled {
    opacity: 0.5;
    pointer-events: none;
    color: hsl(var(--text-muted));
  }

  .sui-toggle-label:active .sui-toggle-switch {
    transform: scale(0.85);
  }

  .sui-toggle-container {
    --toggle-height: 12px;
    --toggle-width: 40px;
    display: flex;
    align-items: center;
    cursor: pointer;
    transition: all .15s ease;
    background-color: hsl(var(--default-base));
    width: var(--toggle-width);
    height: var(--toggle-height);
    border-radius: var(--toggle-height);
  }

  .sui-toggle-switch {
    --switch: calc(var(--toggle-height) * 1.75);
    height: var(--switch);
    width: var(--switch);
    background-color: hsl(var(--text-muted));
    border-radius: var(--toggle-height);
    position: relative;
    left: 0;
    transition: all .15s ease;
    will-change: transform;
  }

  .sui-toggle-container:has(.sui-checkbox:checked) .sui-toggle-switch {
    left: calc(100% - var(--switch));
    background-color: hsl(var(--text-normal));
  }

  .sui-toggle-label.sm .sui-toggle-container {
    --toggle-height: 10px;
    --toggle-width: 32px;
  }

  .sui-toggle-label.sm .sui-toggle-switch {
    --switch: calc(var(--toggle-height) * 1.65);
  }

  .sui-toggle-label.lg .sui-toggle-container {
    --toggle-height: 16px;
    --toggle-width: 48px;
  }

  .sui-toggle-label.lg .sui-toggle-switch {
    --switch: calc(var(--toggle-height) * 1.65);
  }

  .sui-toggle-label.primary .sui-toggle-container:has(.sui-checkbox:checked) {
    background-color: hsl(var(--primary-base));
  }

  .sui-toggle-label.success .sui-toggle-container:has(.sui-checkbox:checked) {
    background-color: hsl(var(--success-base));
  }

  .sui-toggle-label.warning .sui-toggle-container:has(.sui-checkbox:checked) {
    background-color: hsl(var(--warning-base));
  }

  .sui-toggle-label.danger .sui-toggle-container:has(.sui-checkbox:checked) {
    background-color: hsl(var(--danger-base));
  }

  .req-star {
    color: hsl(var(--danger-base));
    font-weight: 700;
  }

  .sui-checkbox {
    width: 0;
    height: 0;
    visibility: hidden;
    opacity: 0;
    margin: 0;
  }
</style>
