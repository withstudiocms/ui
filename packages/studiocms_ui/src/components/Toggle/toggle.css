.sui-toggle-label {
  display: flex;
  flex-direction: row;
  align-items: center;
  gap: .5rem;
  position: relative;
  margin: .25rem 0;
}

.sui-toggle-label.disabled {
  opacity: 0.5;
  pointer-events: none;
  color: hsl(var(--text-muted));
}

.sui-toggle-label:active .sui-toggle-switch {
  transform: scale(0.85);
}

.sui-toggle-container {
  --toggle-height: 12px;
  --toggle-width: 40px;
  display: flex;
  align-items: center;
  cursor: pointer;
  transition: all .15s ease;
  background-color: hsl(var(--default-base));
  width: var(--toggle-width);
  height: var(--toggle-height);
  border-radius: var(--radius-full);
}

.sui-toggle-switch {
  --switch: calc(var(--toggle-height) * 1.75);
  height: var(--switch);
  width: var(--switch);
  background-color: hsl(var(--text-muted));
  border-radius: var(--radius-full);
  position: relative;
  left: 0;
  transition: all .15s ease;
  will-change: transform;
}

.sui-toggle-switch:focus-visible {
  outline: 2px solid hsl(var(--text-normal));
  outline-offset: 2px;
}

.sui-toggle-container:has(.sui-toggle-checkbox:checked) .sui-toggle-switch {
  left: calc(100% - var(--switch));
  background-color: hsl(var(--text-normal));
}

.sui-toggle-label.sm .sui-toggle-container {
  --toggle-height: 10px;
  --toggle-width: 32px;
}

.sui-toggle-label.sm .sui-toggle-switch {
  --switch: calc(var(--toggle-height) * 1.65);
}

.sui-toggle-label.lg .sui-toggle-container {
  --toggle-height: 16px;
  --toggle-width: 48px;
}

.sui-toggle-label.lg .sui-toggle-switch {
  --switch: calc(var(--toggle-height) * 1.65);
}

.sui-toggle-label.primary .sui-toggle-container:has(.sui-toggle-checkbox:checked) {
  background-color: hsl(var(--primary-base));
}

.sui-toggle-label.success .sui-toggle-container:has(.sui-toggle-checkbox:checked) {
  background-color: hsl(var(--success-base));
}

.sui-toggle-label.warning .sui-toggle-container:has(.sui-toggle-checkbox:checked) {
  background-color: hsl(var(--warning-base));
}

.sui-toggle-label.danger .sui-toggle-container:has(.sui-toggle-checkbox:checked) {
  background-color: hsl(var(--danger-base));
}

.sui-toggle-label.info .sui-toggle-container:has(.sui-toggle-checkbox:checked) {
  background-color: hsl(var(--info-base));
}

.sui-toggle-label.mono .sui-toggle-container:has(.sui-toggle-checkbox:checked) {
  background-color: hsl(var(--mono-base));
}

.sui-toggle-label.mono .sui-toggle-switch {
  border: 1px solid hsl(var(--text-inverted));
}

.req-star {
  color: hsl(var(--danger-base));
  font-weight: 700;
}

.sui-toggle-checkbox {
  width: 0;
  height: 0;
  visibility: hidden;
  opacity: 0;
  margin: 0;
}
