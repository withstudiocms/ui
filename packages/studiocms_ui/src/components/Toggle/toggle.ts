function loadToggles() {
	const allToggleElements = document.querySelectorAll<HTMLDivElement>('.sui-toggle-container');
	const allToggles = document.querySelectorAll<HTMLInputElement>('.sui-toggle-checkbox');

	for (const element of allToggleElements) {
		if (element.dataset.initialized) continue;

		element.dataset.initialized = 'true';

		element.addEventListener('keydown', (e) => {
			if (e.key !== 'Enter' && e.key !== ' ') return;

			e.preventDefault();

			const checkbox = element.querySelector<HTMLInputElement>('.sui-toggle-checkbox');

			if (!checkbox) return;

			checkbox.click();
		});
	}

	for (const box of allToggles) {
		if (box.dataset.initialized) continue;

		box.dataset.initialized = 'true';

		box.addEventListener('change', (e) => {
			(box.previousSibling as HTMLDivElement).ariaChecked = (e.target as HTMLInputElement).checked
				? 'true'
				: 'false';
		});
	}
}

document.addEventListener('astro:page-load', loadToggles);
