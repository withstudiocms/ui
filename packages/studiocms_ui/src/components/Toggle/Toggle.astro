---
import type { StudioCMSColorway } from '../../utils/colors.js';
import { generateID } from '../../utils/generateID.js';
import './toggle.css';

/**
 * The props for the toggle component
 */
interface Props {
	/**
	 * The label of the toggle.
	 */
	label: string;
	/**
	 * The size of the toggle. Defaults to `md`.
	 */
	size?: 'sm' | 'md' | 'lg';
	/**
	 * The color of the toggle. Defaults to `default`.
	 */
	color?: StudioCMSColorway;
	/**
	 * Whether the toggle is checked by default. Defaults to `false`.
	 */
	defaultChecked?: boolean;
	/**
	 * Whether the toggle is disabled. Defaults to `false`.
	 */
	disabled?: boolean;
	/**
	 * The name of the toggle.
	 */
	name?: string;
	/**
	 * Whether the toggle is required. Defaults to `false`.
	 */
	isRequired?: boolean;
}

const {
	size = 'md',
	color = 'default',
	defaultChecked,
	disabled,
	name = generateID('checkbox'),
	label,
	isRequired,
} = Astro.props;
---
<label
  class="sui-toggle-label"
  for={name}
  class:list={[
    disabled && "disabled",
    color,
    size,
  ]}
>
  <div class="sui-toggle-container">
    <div 
      class="sui-toggle-switch"
      tabindex="0"
      role="checkbox"
      aria-checked={defaultChecked}
      aria-label={label}
    />
    <input
      type="checkbox"
      name={name}
      id={name}
      checked={defaultChecked}
      disabled={disabled}
      required={isRequired}
      class="sui-toggle-checkbox"
      hidden
    />
  </div>
  <span id={`label-${name}`}>
    {label} <span class="req-star">{isRequired && "*"}</span>
  </span>
</label>
<script>
import "studiocms:ui/scripts/toggle"
</script>
