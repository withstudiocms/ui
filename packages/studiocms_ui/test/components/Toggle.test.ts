import { describe, expect } from 'vitest';
import Toggle from '../../src/components/Toggle/Toggle.astro';
import { test } from '../fixtures/vitest/AstroContainer';

const mockProps = {
	label: 'Toggle Label',
	name: 'toggle',
};

describe('Toggle Component', () => {
	test('renders Toggle component correctly', async ({ renderComponent }) => {
		const result = await renderComponent(Toggle, 'Toggle', { props: mockProps });
		expect(result).toMatchSnapshot();
	});

	test.for([
		{ size: 'sm' },
		{ size: 'md' },
		{ size: 'lg' },
		{ disabled: true },
		{ defaultChecked: true },
		{ isRequired: true },
		{ color: 'primary' },
		{ color: 'default' },
		{ color: 'success' },
		{ color: 'danger' },
		{ color: 'warning' },
		{ color: 'info' },
		{ color: 'mono' },
	])('renders Toggle component with various props: %o', async (props, { renderComponent }) => {
		const result = await renderComponent(Toggle, 'Toggle', { props: { ...mockProps, ...props } });
		expect(result).toMatchSnapshot();
	});
});
