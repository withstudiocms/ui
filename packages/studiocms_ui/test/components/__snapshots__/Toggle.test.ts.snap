// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Toggle Component > renders Toggle component correctly 1`] = `"<label for="toggle" class="sui-toggle-label default md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { color: 'danger' } 1`] = `"<label for="toggle" class="sui-toggle-label danger md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { color: 'default' } 1`] = `"<label for="toggle" class="sui-toggle-label default md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { color: 'info' } 1`] = `"<label for="toggle" class="sui-toggle-label info md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { color: 'mono' } 1`] = `"<label for="toggle" class="sui-toggle-label mono md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { color: 'primary' } 1`] = `"<label for="toggle" class="sui-toggle-label primary md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { color: 'success' } 1`] = `"<label for="toggle" class="sui-toggle-label success md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { color: 'warning' } 1`] = `"<label for="toggle" class="sui-toggle-label warning md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { defaultChecked: true } 1`] = `"<label for="toggle" class="sui-toggle-label default md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="true" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" checked class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { disabled: true } 1`] = `"<label for="toggle" class="sui-toggle-label disabled default md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" disabled class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { isRequired: true } 1`] = `"<label for="toggle" class="sui-toggle-label default md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" required class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star">*</span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { size: 'lg' } 1`] = `"<label for="toggle" class="sui-toggle-label default lg"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { size: 'md' } 1`] = `"<label for="toggle" class="sui-toggle-label default md"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;

exports[`Toggle Component > renders Toggle component with various props: { size: 'sm' } 1`] = `"<label for="toggle" class="sui-toggle-label default sm"> <div class="sui-toggle-container"> <div class="sui-toggle-switch" tabindex="0" role="checkbox" aria-checked="false" aria-label="Toggle Label"></div> <input type="checkbox" name="toggle" id="toggle" class="sui-toggle-checkbox" hidden> </div> <span id="label-toggle"> Toggle Label <span class="req-star"></span> </span> </label> <script type="module" src="/mock/path/Toggle.astro?astro&type=script&index=0&lang.ts"></script>"`;
